### PR TITLE
PR for issue 118 - Requesting help in using the new Classifications programatically

### PR DESCRIPTION
**Pull request by anandbagmar**
This PR contains the following changes:

1. Made class Classification implement the "Serializable" interfact. 
2. When evaluating the macro for Tokens, if we get a ClassCastException, just use the "value". 
3. Fixed a typo in method name - 'evaluaeMacro' 

Also - 
1. This works with Jenkins project type - Pipeline
2. It is backwards compatible (with standalone type jobs)

Will update the issue #118 with details on how I am now invoking this plugin and adding classifications dynamically.

### CODE REVIEW

**Comment by damianszczepanik** [comment]
Thanks for the PR

1. Great - I think it fixed one issue I've recently found: classifications were not populated when the job was copied
2. Tell me more about ``ClassCastException`` - when it happens and can we make some prevention instead of catching such exception from external library?
3. Nice

**Comment by anandbagmar** [comment]
Not sure when, rather why this happens. 1 difference is that I am using the Pipeline plugin (NOT the Freestyle project) in Jenkins to create my jobs (via Jenkins file). As a result, this exception is seen when trying to get the value from the macro

```
java.lang.ClassCastException: org.jenkinsci.plugins.workflow.job.WorkflowRun cannot be cast to hudson.model.AbstractBuild
	at net.masterthought.jenkins.CucumberReportPublisher.evaluaeMacro(CucumberReportPublisher.java:250)
	at net.masterthought.jenkins.CucumberReportPublisher.generateReport(CucumberReportPublisher.java:164)
	at net.masterthought.jenkins.CucumberReportPublisher.perform(CucumberReportPublisher.java:109)
	at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:78)
	at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:65)
	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution$1$1.call(SynchronousNonBlockingStepExecution.java:49)
	at hudson.security.ACL.impersonate(ACL.java:221)
	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution$1.run(SynchronousNonBlockingStepExecution.java:46)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by damianszczepanik** [comment]
I have prepared PR for this problem. Remove this code, rebase with master and try again

**Comment by anandbagmar** [comment]
With locally build hpi from latest master, i do not see the error anymore and am able to see the custom classifications as well in the reports. While you are at it, you may want to rename the evaluaeMacro method as well before you publish the new version of the plugin :) 

Thanks!

**Comment by damianszczepanik** [comment]
Can you resolve conflict and update PR?

**Comment by anandbagmar** [comment]
done. conflicts resolved

File: src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java [hunk 1]
```diff
@@ -241,12 +244,15 @@ private boolean hasReportFailed(Reportable result, TaskListener listener) {
...
     }
 
-    private String evaluaeMacro(Run<?, ?> build, FilePath workspace, TaskListener listener, String value) throws InterruptedException, IOException {
+    private String evaluateMacro(Run<?, ?> build, FilePath workspace, TaskListener listener, String value) throws InterruptedException, IOException {
         try {
             return TokenMacro.expandAll(build, workspace, listener, value);
         } catch (MacroEvaluationException e) {
             log(listener, String.format("Could not evaluate macro '%s': %s", value, e.getMessage()));
             return value;
+        } catch (ClassCastException e) {
```
Comment (damianszczepanik) [inline src/main/java/net/masterthought/jenkins/CucumberReportPublisher.java]: This should be removed

**Comment by anandbagmar** [comment]
done. removed the catch clause which is not required anymore

**Comment by damianszczepanik** [comment]
But still commits are not squashed and also commit message does not refer to the change. Update it please and test with new and existing reports

**Comment by anandbagmar** [comment]
closing this PR. Will raise a fresh PR for the same

**Comment by anandbagmar** [comment]
Done. Raised a new PR #142 for this after cleanup and retesting.



Regards,



Anand Bagmar
about.me/anand.bagmar
[image: Anand Bagmar on about.me]
  <http://about.me/anand.bagmar>

**Pull request closed without merging**